core/blockchain.go

```diff
@@ -885,7 +886,11 @@ func (self *BlockChain) InsertChain(chain types.Blocks) (int, error) {
 
 		// Create a new statedb using the parent block and report an
 		// error if it fails.
-		statedb, err := state.New(self.GetBlock(block.ParentHash()).Root(), self.chainDb)
+		if statedb == nil {
+			statedb, err = state.New(self.GetBlock(block.ParentHash()).Root(), self.chainDb)
+		} else {
+			statedb, err = statedb.New(chain[i-1].Root())
+		}
```

Comment:
Actually, that's what I did first and it fails on out tests, since blocks can contain bad state roots, which it wouldn't detect if I just reused the previous one (i.e. we never validate it currently, just pull it up from th edb).